### ISSUE TITLE
Strange SQL parse failure

### ISSUE THREAD

**Issue by gmr** [lelit/pglast#29]
I've run into a weird parsing error that seems to happen on `pglast.split`.

Given the following code, an exception is raised on L9 in both 1.5 and in master:

```python
import pglast


SQL = 'CREATE CONSTRAINT TRIGGER monitor_account_status AFTER UPDATE ON accounts DEFERRABLE INITIALLY DEFERRED FOR EACH ROW WHEN ((((new.status_id = ANY (ARRAY[4, 5, 6])) <> (old.status_id = ANY (ARRAY[4, 5, 6]))) OR (new.status_id = 7))) EXECUTE PROCEDURE log_account_status_change();'

result = pglast.parse_sql(SQL)
print(result)

for statement in pglast.split(SQL):
    print(statement)
    result = pglast.parse_sql(statement)
    print(result)
```

The output:
```python
[{'RawStmt': {'stmt': {'CreateTrigStmt': {'trigname': 'monitor_account_status', 'relation': {'RangeVar': {'relname': 'accounts', 'inh': True, 'relpersistence': 'p', 'location': 65}}, 'funcname': [{'String': {'str': 'log_account_status_change'}}], 'row': True, 'events': 16, 'whenClause': {'BoolExpr': {'boolop': 1, 'args': [{'A_Expr': {'kind': 0, 'name': [{'String': {'str': '<>'}}], 'lexpr': {'A_Expr': {'kind': 1, 'name': [{'String': {'str': '='}}], 'lexpr': {'ColumnRef': {'fields': [{'String': {'str': 'new'}}, {'String': {'str': 'status_id'}}], 'location': 126}}, 'rexpr': {'A_ArrayExpr': {'elements': [{'A_Const': {'val': {'Integer': {'ival': 4}}, 'location': 153}}, {'A_Const': {'val': {'Integer': {'ival': 5}}, 'location': 156}}, {'A_Const': {'val': {'Integer': {'ival': 6}}, 'location': 159}}], 'location': 147}}, 'location': 140}}, 'rexpr': {'A_Expr': {'kind': 1, 'name': [{'String': {'str': '='}}], 'lexpr': {'ColumnRef': {'fields': [{'String': {'str': 'old'}}, {'String': {'str': 'status_id'}}], 'location': 168}}, 'rexpr': {'A_ArrayExpr': {'elements': [{'A_Const': {'val': {'Integer': {'ival': 4}}, 'location': 195}}, {'A_Const': {'val': {'Integer': {'ival': 5}}, 'location': 198}}, {'A_Const': {'val': {'Integer': {'ival': 6}}, 'location': 201}}], 'location': 189}}, 'location': 182}}, 'location': 164}}, {'A_Expr': {'kind': 0, 'name': [{'String': {'str': '='}}], 'lexpr': {'ColumnRef': {'fields': [{'String': {'str': 'new'}}, {'String': {'str': 'status_id'}}], 'location': 211}}, 'rexpr': {'A_Const': {'val': {'Integer': {'ival': 7}}, 'location': 227}}, 'location': 225}}], 'location': 207}}, 'isconstraint': True, 'deferrable': True, 'initdeferred': True}}, 'stmt_len': 277}}]
```

And the traceback:
```python
Traceback (most recent call last):
  File "/Users/gavinr/Source/Applications/pglifecycle/example.py", line 9, in <module>
    for statement in pglast.split(SQL):
  File "/Users/gavinr/Source/Libraries/pglast/pglast/__init__.py", line 95, in split
    printed_pt = parse_sql(printed)[0]
  File "pglast/parser.pyx", line 76, in pglast.parser.parse_sql
    raise ParseError(message, cursorpos)
```

Any thoughts on what it could be? I'm happy to try and track it down if you can point me in the right direction.

**Comment by lelit** [comment]
"Gavin M. Roy" <notifications@github.com> writes:

> I've run into a weird parsing error that seems to happen on `pglast.split`.

This seems related to a missing () around a term in the condition:
```
$ echo "CREATE CONSTRAINT TRIGGER monitor_account_status AFTER UPDATE ON accounts
DEFERRABLE INITIALLY DEFERRED FOR EACH ROW WHEN ((((new.status_id = ANY (ARRAY[4, 5,
6])) <> (old.status_id = ANY (ARRAY[4, 5, 6]))) OR (new.status_id = 7))) EXECUTE
PROCEDURE log_account_status_change();" | pgpp 
CREATE CONSTRAINT TRIGGER monitor_account_status
  AFTER UPDATE
  ON accounts DEFERRABLE INITIALLY DEFERRED
  FOR EACH ROW
    WHEN (((   (new.status_id = ANY(ARRAY[4, 5, 6]) <> old.status_id = ANY(ARRAY[4, 5, 6]))
            OR (new.status_id = 7))))
    EXECUTE PROCEDURE log_account_status_change()
.../pglast/__init__.py:53: RuntimeWarning: Detected a bug in pglast serialization, please report: syntax error at or near "=", at location 193
```

That "<> old.status_id = ANY(ARRAY[4, 5, 6])" should probably be
"<> (old.status_id = ANY(ARRAY[4, 5, 6]))"

I'm sorry, but I won't be able to switch focus on this immediately.

Thanks&bye, lele.

**Comment by gmr** [comment]
Thanks, but the point was it parses ok with `pglast.sql_parse` but not with `pglast.split`. Also the trigger is valid, runs in 9.0 and 10, and the SQL is the product of pg_dump.

**Comment by lelit** [comment]
"Gavin M. Roy" <notifications@github.com> writes:

> Thanks, but the point was it parses ok with `pglast.sql_parse` but not with
> `pglast.split`.

The split() method, as the pretty_print() one, has a default "safety belt": to
do its job it uses the RawStream printer to print each statement after the
parse, and to be sure that the process did not introduce any error it reparses
the result (see https://github.com/lelit/pglast/blob/master/pglast/__init__.py#L95). 

Given that, as said, there is a bug in handling that boolean expression, the
printer produced an unparsable result.

Will try to investigate as soon as possible, but I'll be very busy for another
couple of days, so do not keep your breath.

**Comment by gmr** [comment]
Ok, thanks

**Comment by lelit** [comment]
This should be fixed in current master, thanks again for the report.

**Comment by lelit** [comment]
Forgot to close, fixed since version 1.6.
